File: 20190919file_benchmarks/benchmark.R
library(fst)
library(microbenchmark)
library(data.table)
library(arrow)
library(feather)
library(stringr)
library(dplyr)

files <- c("2016Q4", "yellow_tripdata_2010-01")
names <- c("fanniemae", "nyctaxi")
seps <- c("|", ",")

create_files <- function(base) {
  df <- arrow::read_parquet(str_c(base, "_snappy.parquet"))
  feather::write_feather(df, str_c(base, "_v1.feather"))
  fst::write_fst(df, str_c(base, "_0.fst"), compress=0)
  fst::write_fst(df, str_c(base, "_50.fst"), compress=50)
  saveRDS(df, str_c(base, "_compressed.rds"), compress=TRUE)
  saveRDS(df, str_c(base, "_uncompressed.rds"), compress=FALSE)
}

do_benchmark <- function(index) {
  base <- files[index]
  sep <- seps[index]

  csv_path <- str_c("data/", base, ".csv")
  feather_v1_path <- str_c(base, "_v1.feather")
  feather_unc_path <- str_c(base, "_uncompressed.feather")
  feather_lz4_path <- str_c(base, "_lz4.feather")
  feather_zstd_path <- str_c(base, "_zstd.feather")
  fst_0_path <- str_c(base, "_0.fst")
  fst_50_path <- str_c(base, "_50.fst")
  parquet_unc_path <- str_c(base, "_uncompressed.parquet")
  parquet_snappy_path <- str_c(base, "_snappy.parquet")
  rds_unc_path <- str_c(base, "_uncompressed.rds")
  rds_compressed_path <- str_c(base, "_compressed.rds")

  mbm <- microbenchmark(
     csv_fread=data.table::fread(csv_path, sep=sep, header=FALSE),
     fst_unc=fst::read_fst(fst_0_path),
     fst_50=fst::read_fst(fst_50_path),
     feather_v1=feather::read_feather(feather_v1_path),
     feather_unc=arrow::read_feather(feather_unc_path),
     feather_lz4=arrow::read_feather(feather_lz4_path),
     feather_zstd=arrow::read_feather(feather_zstd_path),
     parquet_unc=arrow::read_parquet(parquet_unc_path),
     parquet_snappy=arrow::read_parquet(parquet_snappy_path),
     rds_unc=readRDS(rds_unc_path),
     rds_compressed=readRDS(rds_compressed_path),
     times=5
  )
  mbm <- data.frame(mbm) %>% dplyr::group_by(expr) %>% dplyr::summarize(time=mean(time))
  mbm$dataset <- names[index]
  mbm
}

do_write_benchmark <- function(index) {
  base <- files[index]
  sep <- seps[index]

  df <- arrow::read_parquet(str_c(base, "_snappy.parquet"))

  mbm <- microbenchmark(
     fst_unc=fst::write_fst(df, str_c(base, "_0.fst"), compress=0),
     fst_50=fst::write_fst(df, str_c(base, "_50.fst"), compress=50),
     feather_v1=feather::write_feather(df, str_c(base, "_v1.feather")),
     feather_unc=arrow::write_feather(df, str_c(base, "_unc_r.feather"),
          compression="uncompressed"),
     feather_lz4=arrow::write_parquet(df, str_c(base, "_lz4_r.feather"),
          compression="lz4"),
     feather_zstd=arrow::write_parquet(df, str_c(base, "_zstd_r.feather"),
          compression="zstd"),
     parquet_unc=arrow::write_parquet(df, str_c(base, "_unc_r.parquet"),
          compression="uncompressed"),
     parquet_snappy=arrow::write_parquet(df, str_c(base, "_snappy_r.parquet"),
          compression="snappy"),
     rds_compressed=saveRDS(df, str_c(base, "_compressed.rds"), compress=TRUE),
     rds_unc=saveRDS(df, str_c(base, "_uncompressed.rds"), compress=FALSE),
     times=1
  )
  mbm <- data.frame(mbm) %>% dplyr::group_by(expr) %>% dplyr::summarize(time=mean(time))
  mbm$dataset <- names[index]
  mbm
}

generate_files <- function() {
  for (base in files) {
    create_files(base)
  }
}

# generate_files()

print(str_c("Using ", arrow::cpu_count(), " threads"))

results <- dplyr::bind_rows(do_benchmark(1), do_benchmark(2))
print(results)
write.csv(results, str_c("r_read_results_", arrow::cpu_count(), ".csv"))

write_results <- dplyr::bind_rows(do_write_benchmark(1), do_write_benchmark(2))
print(write_results)
write.csv(write_results, str_c("r_write_results_", arrow::cpu_count(), ".csv"))

File: 20190919file_benchmarks/benchmark.py
# flake8: noqa

import pyarrow.feather as feather
import pandas as pd
import json
import numpy as np
import pyarrow as pa
import pyarrow.parquet as pq
from pandas.util.testing import rands
import gc
import time


def get_timing(f, niter):
    start = time.clock_gettime(time.CLOCK_REALTIME)
    for i in range(niter):
        f()
    result = (time.clock_gettime(time.CLOCK_REALTIME) - start) / niter
    return result


class Benchmarker:

    def __init__(self, file_info):
        self.base = file_info['base']
        (self.csv_path,
         self.sep,
         self.header) = unpack(file_info['source'], 'path', 'sep', 'header')

        self.parquet_unc_path = '{}_uncompressed.parquet'.format(self.base)
        self.parquet_snappy_path = '{}_snappy.parquet'.format(self.base)
        self.feather_unc_path = '{}_uncompressed.feather'.format(self.base)
        self.feather_lz4_path = '{}_lz4.feather'.format(self.base)
        self.feather_zstd_path = '{}_zstd.feather'.format(self.base)

    def bench_read(self, niter=5):
        cases = [
            ('parquet (UNC)', 'arrow Table',
             lambda: pq.read_table(self.parquet_unc_path, memory_map=False)),
            ('parquet (UNC)', 'pandas',
             lambda: (pq.read_table(self.parquet_unc_path, memory_map=False)
                      .to_pandas())),
            ('parquet (SNAPPY)', 'arrow Table',
             lambda: pq.read_table(self.parquet_snappy_path,
                                   memory_map=False)),
            ('parquet (SNAPPY)', 'pandas',
             lambda: (pq.read_table(self.parquet_snappy_path, memory_map=False)
                      .to_pandas())),
            ('feather V2 (UNC)', 'pandas',
             lambda: feather.read_feather(self.feather_unc_path,
                                          memory_map=False)),
            ('feather V2 (LZ4)', 'pandas',
             lambda: feather.read_feather(self.feather_lz4_path,
                                          memory_map=False)),
            ('feather V2 (ZSTD)', 'pandas',
             lambda: feather.read_feather(self.feather_zstd_path,
                                          memory_map=False)),
            ('feather V2 (UNC)', 'arrow Table',
             lambda: feather.read_table(self.feather_unc_path,
                                        memory_map=False)),
            ('feather V2 (LZ4)', 'arrow Table',
             lambda: feather.read_table(self.feather_lz4_path,
                                        memory_map=False)),
            ('feather V2 (ZSTD)', 'arrow Table',
             lambda: feather.read_table(self.feather_zstd_path,
                                        memory_map=False)),
        ]

        return self._bench_cases(cases, niter)

    def bench_write(self, niter=2):
        print("Reading text file: {}".format(self.csv_path))
        df = pd.read_csv(self.csv_path, sep=self.sep, header=self.header,
                         low_memory=False)
        if self.header is None:
            df.columns = ['f{}'.format(i) for i in range(len(df.columns))]

        def _get_table(df):
            return (pa.Table.from_pandas(df, preserve_index=False)
                    .replace_schema_metadata(None))

        t = _get_table(df)

        cases = [
            ('parquet (UNC)', 'arrow Table',
             lambda: pq.write_table(t, self.parquet_unc_path,
                                    compression='NONE')),
            ('parquet (UNC)', 'pandas',
             lambda: pq.write_table(_get_table(df), self.parquet_unc_path,
                                    compression='NONE')),
            ('parquet (SNAPPY)', 'arrow Table',
             lambda: pq.write_table(t, self.parquet_snappy_path)),
            ('parquet (SNAPPY)', 'pandas',
             lambda: pq.write_table(_get_table(df), self.parquet_snappy_path)),
            ('feather V2 (UNC)', 'pandas',
             lambda: feather.write_feather(df, self.feather_unc_path,
                                           compression='uncompressed')),
            ('feather V2 (UNC)', 'arrow Table',
             lambda: feather.write_feather(t, self.feather_unc_path,
                                           compression='uncompressed')),
            ('feather V2 (LZ4)', 'pandas',
             lambda: feather.write_feather(df, self.feather_lz4_path,
                                           compression='lz4')),
            ('feather V2 (LZ4)', 'arrow Table',
             lambda: feather.write_feather(t, self.feather_lz4_path,
                                           compression='lz4')),
            ('feather V2 (ZSTD)', 'pandas',
             lambda: feather.write_feather(df, self.feather_zstd_path,
                                           compression='zstd')),
            ('feather V2 (ZSTD)', 'arrow Table',
             lambda: feather.write_feather(t, self.feather_zstd_path,
                                           compression='zstd'))
        ]

        return self._bench_cases(cases, niter)

    def _bench_cases(self, cases, niter):
        results = []
        for name, output_type, f in cases:
            print(name)
            result = (name, output_type, get_timing(f, niter))
            print(result)
            results.append(result)
        return pd.DataFrame.from_records(results,
                                         columns=['expr', 'output_type',
                                                  'mean'])


def unpack(d, *fields):
    return (d[f] for f in fields)



files = {
    'fanniemae': {
        'base': '2016Q4',
        'source': {
            'path': 'data/2016Q4.csv',
            'sep': '|',
            'header': None
        }
    },
    'nyctaxi': {
        'base': 'yellow_tripdata_2010-01',
        'source': {
            'path': 'data/yellow_tripdata_2010-01.csv',
            'sep': ',',
            'header': 0
        }
    }
}


def run_benchmarks(num_threads, what='read'):
    pa.set_cpu_count(num_threads)

    all_results = []
    for name, info in files.items():
        benchmarker = Benchmarker(info)
        if what == 'read':
            print("Benchmarking reads")
            file_results = benchmarker.bench_read()
        elif what == 'write':
            print("Benchmarking writes")
            file_results = benchmarker.bench_write()
        else:
            raise ValueError(what)
        file_results['dataset'] = name
        all_results.append(file_results)

    print(all_results)
    return pd.concat(all_results, ignore_index=True)



# for i in range(5):
#     pq.read_table('yellow_tripdata_2010-01.parquet').to_pandas()

# write_files(files)

num_threads_cases = [1, 4]

for nthreads in num_threads_cases:
    write_results = run_benchmarks(nthreads, what='write')
    write_results.to_csv('py_write_results_{}.csv'.format(nthreads))

    read_results = run_benchmarks(nthreads, what='read')
    read_results.to_csv('py_read_results_{}.csv'.format(nthreads))

# for nthreads in num_threads_cases:
#     run_benchmarks(nthreads)

# ('pyarrow.parquet', 1.5470361709594727)
# ('pyarrow.parquet-pandas', 2.925654172897339)
# ('pyarrow.feather', 1.6384665012359618)
